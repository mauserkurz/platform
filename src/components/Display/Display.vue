<template>
  <div class="display">
    <canvas ref="canvas"></canvas>
    <img
      class="display__sprite"
      ref="otherSprites"
      src="@/assets/sprites.png"/>
    <img
      class="display__sprite"
      ref="playerSprites"
      src="@/assets/player.png"/>
  </div>
</template>

<script>
import { mapState, mapGetters } from 'vuex';
import {
  SCALE,
  PLAYER_CONFIG,
  LAVA_CONFIG,
  COIN_CONFIG,
  EMPTY_CONFIG,
  STATUS_MAP,
  MIN_CANVAS_SIZE,
} from '@/consts';

export default {
  name: 'Display',

  data() {
    return {
      cx: null,
      flipPlayer: false,
      viewport: {
        left: 0,
        top: 0,
        width: 0,
        height: 0,
      },
    };
  },

  watch: {
    status: 'init',
    stateChange: 'syncState',
    countEndLevel: 'clear',
  },

  computed: {
    ...mapState('core', ['actors', 'status']),
    ...mapState('level', ['levelWidth', 'levelHeight', 'rows', 'countEndLevel']),
    ...mapGetters('core', ['player', 'stateChange']),
  },

  methods: {
    /**
     * Развернет спрайт в противоположное напрваление по горизонтали
     * @param {number} around - положение спрайта
     */
    flipHorizontally(around) {
      this.cx.translate(around, 0);
      this.cx.scale(-1, 1);
      this.cx.translate(-around, 0);
    },

    /**
     * Обновит видимую часть уровня на canvas, при этом не двигаясь когда
     * игрок сместится больше чем на 1/3 ширины видимой части
     * @param {number} levelWidth - ширина уровня
     * @param {number} levelHeight - высота уровня
     * @param {object} player - данные о игроке
     */
    updateViewport(levelWidth, levelHeight, player) {
      const {
        left,
        top,
        width,
        height,
      } = this.viewport;
      const margin = width / 3;
      const center = player.pos.plus(player.size.times(0.5));

      if (center.x < left + margin) {
        this.viewport.left = Math.max(center.x - margin, 0);
      } else if (center.x > left + width - margin) {
        this.viewport.left = Math.min(center.x + margin - width, levelWidth - width);
      }
      if (center.y < top + margin) {
        this.viewport.top = Math.max(center.y - margin, 0);
      } else if (center.y > top + height - margin) {
        this.viewport.top = Math.min(center.y + margin - height, levelHeight - height);
      }
    },

    /**
     * Отрисует фон уровня на canvas состоящий из неподвижных элементов этого уровня
     * @param {array} rows - двумерный массив сущностей находящихся на уровне, предствленных
     * строками типов сущностей или конструкторами сущностей
     */
    drawBackground(rows) {
      const {
        left,
        top,
        width,
        height,
      } = this.viewport;
      const xStart = Math.floor(left);
      const xEnd = Math.ceil(left + width);
      const yStart = Math.floor(top);
      const yEnd = Math.ceil(top + height);

      for (let y = yStart; y < yEnd; y += 1) {
        for (let x = xStart; x < xEnd; x += 1) {
          const tile = rows[y][x];

          if (tile === EMPTY_CONFIG.TYPE) {
            // eslint-disable-next-line no-continue
            continue;
          }
          const screenX = (x - left) * SCALE;
          const screenY = (y - top) * SCALE;
          const tileX = tile === LAVA_CONFIG.TYPE ? SCALE : 0;

          this.cx.drawImage(
            this.$refs.otherSprites,
            tileX,
            0,
            SCALE,
            SCALE,
            screenX,
            screenY,
            SCALE,
            SCALE,
          );
        }
      }
    },

    /**
     * Обновит фон в зависимости от статуса игры
     * @param {string} status - статус уровня, который может быть - победа, поражение или играет
     */
    fillDisplay(status) {
      const BACKGROUND_COLOR_MAP = {
        [STATUS_MAP.PLAYING]: '#34a6fb',
        [STATUS_MAP.LOST]: '#2c88d6',
        [STATUS_MAP.WON]: '#44bfff',
      };

      this.cx.fillStyle = BACKGROUND_COLOR_MAP[status];
      this.cx.fillRect(0, 0, this.$refs.canvas.width, this.$refs.canvas.height);
    },

    /**
     * Обновит спрайт игрока на canvas
     * @param {object} player - данные о игроке
     * @param {number} x - ордината игрока в видимой части экрана
     * @param {number} y - абсцисса игрока в видимой части экрана
     * @param {number} width - ширина игрока
     * @param {number} height - высота игрока
     */
    drawPlayer(player, x, y, width, height) {
      const widthWithOverlap = width + PLAYER_CONFIG.X_OVERLAP * 2;
      const xWithoutOverlap = x - PLAYER_CONFIG.X_OVERLAP;

      if (player.speed.x !== 0) {
        this.flipPlayer = player.speed.x < 0;
      }

      let tile = 8;
      if (player.speed.y !== 0) {
        tile = 9;
      } else if (player.speed.x !== 0) {
        tile = Math.floor(Date.now() / 60) % 8;
      }

      this.cx.save();
      if (this.flipPlayer) {
        this.flipHorizontally(xWithoutOverlap + widthWithOverlap / 2);
      }
      const tileX = tile * widthWithOverlap;

      this.cx.drawImage(
        this.$refs.playerSprites,
        tileX,
        0,
        widthWithOverlap,
        height,
        xWithoutOverlap,
        y,
        widthWithOverlap,
        height,
      );
      this.cx.restore();
    },

    /**
     * Отрисует спрайты всех подвижных сущностей на canvas
     * @param {array} actors - список всех подвижных сущностей на уровне
     */
    drawActors(actors) {
      for (const actor of actors) {
        const width = actor.size.x * SCALE;
        const height = actor.size.y * SCALE;
        const x = (actor.pos.x - this.viewport.left) * SCALE;
        const y = (actor.pos.y - this.viewport.top) * SCALE;

        if (actor.type === PLAYER_CONFIG.TYPE) {
          this.drawPlayer(actor, x, y, width, height);
        } else {
          const tileX = (actor.type === COIN_CONFIG.TYPE ? 2 : 1) * SCALE;
          this.cx.drawImage(this.$refs.otherSprites, tileX, 0, width, height, x, y, width, height);
        }
      }
    },

    /**
     * Обновит canvas перерисовав фон, сущносей на экране и их положение в зависимости от
     * данных о уровне для этого кадра анимации
     */
    syncState() {
      if (this.status) {
        this.updateViewport(this.levelWidth, this.levelHeight, this.player);
        this.fillDisplay(this.status);
        this.drawBackground(this.rows);
        this.drawActors(this.actors);
      }
    },

    /**
     * Очистит canvas
     */
    clear() {
      if (this.status) {
        this.cx.clearRect(0, 0, this.$refs.canvas.width, this.$refs.canvas.height);
      }
    },

    /**
     * Подготовит canvas к отрисовке уровня на нем, задав ему размеры и записав контекст
     */
    init() {
      this.$refs.canvas.width = Math.min(MIN_CANVAS_SIZE.WIDTH, this.levelWidth * SCALE);
      this.$refs.canvas.height = Math.min(MIN_CANVAS_SIZE.HEIGHT, this.levelHeight * SCALE);
      this.cx = this.$refs.canvas.getContext('2d');
      this.viewport.width = this.$refs.canvas.width / SCALE;
      this.viewport.height = this.$refs.canvas.height / SCALE;
    },
  },
};
</script>

<style scoped lang="less">
.display {
  &__sprite {
    display: none;
  }
}
</style>
