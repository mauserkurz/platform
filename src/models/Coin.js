import Vector from '@/models/Vector';
import { COIN_CONFIG, STATUS_MAP } from '@/consts';

// TODO DI of Vector
const COIN_SIZE = new Vector(COIN_CONFIG.WIDTH, COIN_CONFIG.HEIGHT);

// TODO add comments
export default class Coin {
  #wobbleSpeed = COIN_CONFIG.WOBBLE_SPEED;

  #wobbleDist = COIN_CONFIG.WOBBLE_DIST;

  #type = COIN_CONFIG.TYPE;

  #size = COIN_SIZE;

  constructor(pos, basePos, wobble) {
    this.pos = pos;
    this.basePos = basePos;
    this.wobble = wobble;
  }

  get size() {
    return this.#size;
  }

  get type() {
    return this.#type;
  }

  update(time) {
    const wobble = this.wobble + time * this.#wobbleSpeed;
    const wobblePos = Math.sin(wobble) * this.#wobbleDist;

    return new Coin(
      this.basePos.plus(new Vector(0, wobblePos)),
      this.basePos, wobble,
    );
  }

  collide({ actors, status }) {
    const filtered = actors.filter(a => a !== this);
    let currentStatus = status;

    if (!filtered.some(a => a.type === COIN_CONFIG.TYPE)) {
      currentStatus = STATUS_MAP.WON;
    }
    return { actors: filtered, status: currentStatus };
  }

  static create(pos) {
    const basePosition = pos.plus(new Vector(
      COIN_CONFIG.POSITION_CORRECTION.X,
      COIN_CONFIG.POSITION_CORRECTION.Y,
    ));

    return new Coin(basePosition, basePosition, Math.random() * Math.PI * 2);
  }
}
