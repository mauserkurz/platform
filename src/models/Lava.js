import Vector from '@/models/Vector';
import { LAVA_CONFIG, WALL_CONFIG, STATUS_MAP } from '@/consts';

// TODO DI of Vector?
const LAVA_SIZE = new Vector(LAVA_CONFIG.WIDTH, LAVA_CONFIG.HEIGHT);

// TODO add comments
export default class Lava {
  #type = LAVA_CONFIG.TYPE;

  #size = LAVA_SIZE;

  constructor(pos, speed, reset) {
    this.pos = pos;
    this.speed = speed;
    this.reset = reset;
  }

  get size() {
    return this.#size;
  }

  get type() {
    return this.#type;
  }

  update(time, touches) {
    const newPos = this.pos.plus(this.speed.times(time));

    if (!touches(newPos, this.size, WALL_CONFIG.TYPE)) {
      return new Lava(newPos, this.speed, this.reset);
    }
    if (this.reset) {
      return new Lava(this.reset, this.speed, this.reset);
    }
    return new Lava(this.pos, this.speed.times(-1));
  }

  // eslint-disable-next-line class-methods-use-this
  collide({ actors }) {
    return { actors, status: STATUS_MAP.LOST };
  }

  static create(pos, ch) {
    const typeMap = {
      '=': new Lava(pos, new Vector(LAVA_CONFIG.HORIZONTAL.X_SPEED, LAVA_CONFIG.HORIZONTAL.Y_SPEED)),
      '|': new Lava(pos, new Vector(LAVA_CONFIG.VERTICAL.X_SPEED, LAVA_CONFIG.VERTICAL.Y_SPEED)),
      v: new Lava(
        pos,
        new Vector(LAVA_CONFIG.FALLING.X_SPEED, LAVA_CONFIG.FALLING.Y_SPEED),
        pos,
      ),
    };

    return typeMap[ch];
  }
}
