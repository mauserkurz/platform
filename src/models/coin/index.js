/**
 * @file фабрика для создания сущности монеты
 */
import createVector from '@/models/vector/';
import { COIN_CONFIG, STATUS_MAP } from '@/consts';

const COIN_SIZE = createVector(COIN_CONFIG.WIDTH, COIN_CONFIG.HEIGHT);

/**
 * создаст новую монету
 * @param {object} pos - новые координаты положения
 * @param {object} basePos - начальные координаты положения
 * @param {number} wobble - коофициент отклонения монеты от центра
 * @return {object} -
 */
export const createCoin = (pos, basePos, wobble) => ({
  pos,
  basePos,
  wobble,

  get size() {
    return COIN_SIZE;
  },

  get type() {
    return COIN_CONFIG.TYPE;
  },

  /**
   * пересоздаст монету сместив её от центра в зависимости от времени анимации (time)
   * @param {number} time - время между новым и прошлым кадром анимации в секундах
   * @return {object} - новая монета
   */
  update(time) {
    const updatedWobble = this.wobble + time * COIN_CONFIG.WOBBLE_SPEED;
    const wobblePos = Math.sin(wobble) * COIN_CONFIG.WOBBLE_DIST;

    return createCoin(
      this.basePos.plus(createVector(0, wobblePos)),
      this.basePos,
      updatedWobble,
    );
  },

  /**
   * удалит монету из уровня и если больше не останется монет, статус игры сменится на победу,
   * вызвается в случае столкновения игрока с монетой
   * @param {array} actors - подвижные объекты на уровне
   * @param {string} status - статус состояния игры: продолжается, победа или поражение
   * @return {{actors: {array}, status: (string)}} - объект с обновленными данными
   * где actors - обновленными данные о подвижных объектах на уровне,
   * status - обновленный статус игры
   */
  collide({ actors, status }) {
    const filtered = actors.filter(a => a !== this);
    let currentStatus = status;

    if (!filtered.some(a => a.type === COIN_CONFIG.TYPE)) {
      currentStatus = STATUS_MAP.WON;
    }
    return {
      actors: filtered,
      status: currentStatus,
    };
  },
});

/**
 * создаст монету для нового уровня со параметрами по умолчанию
 * @param {object} pos - координаты положения
 * @return {object} - новая монета
 */
export default (pos) => {
  const basePosition = pos.plus(createVector(
    COIN_CONFIG.POSITION_CORRECTION.X,
    COIN_CONFIG.POSITION_CORRECTION.Y,
  ));

  return createCoin(basePosition, basePosition, Math.random() * Math.PI * 2);
};
