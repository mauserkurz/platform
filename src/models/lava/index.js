/**
 * @file фабрика для создания сущности лавы
 */

import createVector from '@/models/vector/';
import { LAVA_CONFIG, WALL_CONFIG, STATUS_MAP } from '@/consts';

const LAVA_SIZE = createVector(LAVA_CONFIG.WIDTH, LAVA_CONFIG.HEIGHT);

/**
 * создаст новую сущность лавы
 * @param {object} pos - данные с координатами положения
 * @param {object} speed - данные с вектором, прибавляемыми за 1 секунду
 * @param {object} reset - данные с координатами исходного положения, требуется восстановления
 * положения "капающей" лавы
 * @return {object} - новая сущность лавы
 */
const createLava = (pos, speed, reset) => ({
  pos,
  speed,
  reset,

  get size() {
    return LAVA_SIZE;
  },

  get type() {
    return LAVA_CONFIG.TYPE;
  },

  /**
   * пересоздаст лаву, сместив её координаты если она не коснется стены,
   * в противном случае, если нет данных с исходным положением (reset),
   * то начать двигаться в противоположном направлении от исходного, если есть
   * сбросить положение до исходного
   * @param {number} time - время между новым и прошлым кадром анимации в секундах
   * @param {function} touches - функция, которая вернет булево значение равное
   * true если лава будет пересекать указаный тип объекта уровня, необходима для
   * отскакивания движущейся лавы от стен
   * @return {object} - новая сущность лавы
   */
  update(time, touches) {
    const newPos = this.pos.plus(this.speed.times(time));

    if (!touches(newPos, this.size, WALL_CONFIG.TYPE)) {
      return createLava(newPos, this.speed, this.reset);
    }
    if (this.reset) {
      return createLava(this.reset, this.speed, this.reset);
    }
    return createLava(this.pos, this.speed.times(-1), null);
  },

  /**
   * заменит статус игры на проигрыш, когда будет вызвана при столкновении с игроком
   * @param {array} actors - подвижные объекты на уровне
   * @return {{actors: {array}, status: (string)}} - объект с обновленными данными
   * где actors - обновленными данные о подвижных объектах на уровне,
   * status - обновленный статус игры
   */
  collide({ actors }) {
    return { actors, status: STATUS_MAP.LOST };
  },
});

/**
 * создаст лаву для нового уровня с параметрами по умолчанию, в зависимости от типа лавы
 * @param {object} pos - данные с координатами положения
 * @param {string} char - символ, обозначающий лаву в шаблоне уровня
 * @return {object} - новая сущность лавы
 */
export default (pos, char) => {
  const typeMap = {
    '=': createLava(pos, createVector(LAVA_CONFIG.HORIZONTAL.X_SPEED, LAVA_CONFIG.HORIZONTAL.Y_SPEED)),
    '|': createLava(pos, createVector(LAVA_CONFIG.VERTICAL.X_SPEED, LAVA_CONFIG.VERTICAL.Y_SPEED)),
    v: createLava(
      pos,
      createVector(LAVA_CONFIG.FALLING.X_SPEED, LAVA_CONFIG.FALLING.Y_SPEED),
      pos,
    ),
  };

  return typeMap[char];
};
