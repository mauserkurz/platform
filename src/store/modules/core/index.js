import { setProp } from '@/store/reusable';
import { PLAYER_CONFIG, LAVA_CONFIG, STATUS_MAP } from '@/consts';

/**
 * Вернет значение отвечающее на вопрос пересекает ли сущность actor1 другую сущность actor2
 * @param {object} actor1 - сущность, пересечение которой мы ищем
 * @param {object} actor2 - сущность с которой может пересечься actor1
 * @return {boolean} - значение пересечения
 */
const overlap = (actor1, actor2) => actor1.pos.x + actor1.size.x > actor2.pos.x
  && actor1.pos.x < actor2.pos.x + actor2.size.x
  && actor1.pos.y + actor1.size.y > actor2.pos.y
  && actor1.pos.y < actor2.pos.y + actor2.size.y;

export default {
  namespaced: true,

  state: {
    actors: [],
    status: '',
  },

  getters: {
    /**
     * Найдет игрока среди подвижных сущностей на текущем уровне
     * @param {object} state - состояние стора
     * @return {object} - сущность игрока
     */
    player(state) {
      return state.actors.find(a => a.type === PLAYER_CONFIG.TYPE);
    },

    /**
     * Вернет данные текущего уровня, изменения которых можно отслеживать для
     * @param {array} actors - подвижные объекты на уровне
     * @param {string} status - статус уровня, который может быть - победа, поражение или играет
     * последующей перерисовки во view
     * @return {array} - данные текущего уровня
     */
    stateChange({ actors, status }) {
      return [actors, status];
    },
  },

  mutations: {
    /**
     * Сохранит список подвижных сущностей на уровне
     */
    SET_ACTORS: setProp('actors'),

    /**
     * Сохранит статус игры
     */
    SET_STATUS: setProp('status'),
  },

  actions: {
    /**
     * Обновит состояние уровня в завистимости от действий подвижных сущностей
     * и действий игрока
     * @param {object} context - содержит данные стора и функции для управления им
     * @param {number} time - время в секундах, показываеющее разницу между прошлым
     * и следующим кадрами анимации
     * @param {object} keys - map в котором указаны нажатые клавиши
     * @return {string} - статус игры после обновления состояния уровня
     */
    update(context, { time, keys }) {
      const {
        state,
        getters,
        rootGetters,
        commit,
      } = context;
      const touches = rootGetters['level/touches'];
      const actors = state.actors
        .map(actor => actor.update(time, touches, keys));

      commit('SET_ACTORS', actors);
      if (state.status !== STATUS_MAP.PLAYING) {
        return state.status;
      }
      const { player } = getters;

      if (touches(player.pos, player.size, LAVA_CONFIG.TYPE)) {
        commit('SET_STATUS', STATUS_MAP.LOST);
        return state.status;
      }

      for (const actor of actors) {
        if (actor !== player && overlap(actor, player)) {
          const newState = actor.collide({ actors, status: state.status });

          commit('SET_ACTORS', newState.actors);
          commit('SET_STATUS', newState.status);
        }
      }
      return state.status;
    },
  },
};
