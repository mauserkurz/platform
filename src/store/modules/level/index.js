import { setProp, plus } from '@/store/reusable';
import createVector from '@/models/vector/';
import createPlayer from '@/models/player/';
import createLava from '@/models/lava/';
import createInitialCoin from '@/models/coin/';
import {
  LAVA_CONFIG,
  WALL_CONFIG,
  EMPTY_CONFIG,
  STATUS_MAP,
} from '@/consts';

const LEVEL_CHAR_MAP = {
  '.': EMPTY_CONFIG.TYPE,
  '#': WALL_CONFIG.TYPE,
  '+': LAVA_CONFIG.TYPE,
  '=': createLava,
  '|': createLava,
  v: createLava,
  o: createInitialCoin,
  '@': createPlayer,
};

export default {
  namespaced: true,

  state: {
    levelHeight: 0,
    levelWidth: 0,
    countEndLevel: 0,
    rows: [[]],
  },

  getters: {
    /**
     * Создаст функцию, чтобы можно было проверить пересечение сущности с краем
     * уровня или другой сущностью переданного типа (type), применяется для обработки
     * движения и с учетом столкновения с краем уровня или стеной
     * @param {object} state - состояние стора
     * @return {function} - функция для проверки пересечения, где аргументы
     * pos - позиция сущности,
     * size - размер сущности
     * type - тип сущности, пересечение с которой требуется найти
     * вернет булево значение, при этом true - когда сущность будет пересекать
     * с краем уровня или другой сущностью переданного типа (type)
     */
    touches: state => (pos, size, type) => {
      const xStart = Math.floor(pos.x);
      const xEnd = Math.ceil(pos.x + size.x);
      const yStart = Math.floor(pos.y);
      const yEnd = Math.ceil(pos.y + size.y);

      for (let y = yStart; y < yEnd; y += 1) {
        for (let x = xStart; x < xEnd; x += 1) {
          const isOutside = x < 0 || x >= state.levelWidth
            || y < 0 || y >= state.levelHeight;
          const here = isOutside ? WALL_CONFIG.TYPE : state.rows[y][x];

          if (here === type) {
            return true;
          }
        }
      }
      return false;
    },
  },

  mutations: {
    /**
     * Сохранит высоту уровня
     */
    SET_LEVEL_HEIGHT: setProp('levelHeight'),

    /**
     * Сохранит ширину уровня
     */
    SET_LEVEL_WIDTH: setProp('levelWidth'),

    /**
     * Сохранит двумерный массив неподвижных сущностей на уровне
     */
    SET_ROWS: setProp('rows'),

    /**
     * Добавит 1 к счетчику начала уровня, нужно для запуска уровня в компоненте
     * при смене значения счетчика
     */
    ADD_END_LEVEL: plus('countEndLevel'),
  },

  actions: {
    /**
     * Сгенерирует данные уровня на основе переданного шаблона (plan) и сохранит их в state,
     * сгенерированы и сохранены будут ширина и высота уровня,
     * данные сущностей на уровне и статус игры
     * @param {function} commit - запустит мутацию с перереданными параметрами
     * @param {string} plan
     */
    start({ commit }, plan) {
      const startActors = [];
      const rows = plan
        .trim()
        .split('\n')
        .map(row => row.trim().split(''))
        .map((row, y) => row.map((char, x) => {
          const entity = LEVEL_CHAR_MAP[char];

          if (typeof entity === 'string') {
            return entity;
          }

          startActors.push(entity(createVector(x, y), char));
          return EMPTY_CONFIG.TYPE;
        }));

      commit('SET_LEVEL_HEIGHT', rows.length);
      commit('SET_LEVEL_WIDTH', rows[0].length);
      commit('SET_ROWS', rows);
      commit('core/SET_ACTORS', startActors, { root: true });
      commit('core/SET_STATUS', STATUS_MAP.PLAYING, { root: true });
    },
  },
};
