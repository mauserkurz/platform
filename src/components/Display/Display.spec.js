import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import 'jest-canvas-mock';
import clone from 'ramda/src/clone';
import core from '@/store/modules/core/';
import level from '@/store/modules/level/';
import createPlayer from '@/models/player';
import createVector from '@/models/vector';
import createLava from '@/models/lava';
import { createCoin } from '@/models/coin/';
import Display from './Display.vue';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('component Display', () => {
  const createWrapper = ({ options = {}, modules = {}, needsRendering = false } = {}) => {
    const store = new Vuex.Store({
      namespaced: true,
      modules: Object.assign({}, { core, level }, modules),
    });
    const defaultOptions = { localVue, store };
    const customMount = needsRendering ? mount : shallowMount;
    const wrapper = customMount(Display, Object.assign({}, defaultOptions, options));

    return { store, wrapper };
  };
  // eslint-disable-next-line no-underscore-dangle
  const checkCanvas = wrapper => expect(wrapper.vm.cx.__getEvents()).toMatchSnapshot();

  Date.now = jest.fn().mockReturnValue(1577416453535);
  describe('rendering', () => {
    it('should render html', () => {
      const { wrapper } = createWrapper();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('store interaction', () => {
    it(`при изменении статуса игры должен выставить корректно
      ширину и высоту canvas`, () => {
      const levelClone = clone(level);
      const coreClone = clone(core);

      levelClone.state.levelHeight = 3;
      levelClone.state.levelWidth = 4;
      levelClone.state.rows = [
        ['empty', 'empty', 'empty', 'empty'],
        ['empty', 'empty', 'empty', 'wall'],
        ['empty', 'wall', 'wall', 'wall'],
      ];
      coreClone.state.actors = [
        createLava(createVector(0, 0), '='),
        createPlayer(createVector(2, 1)),
      ];
      const { wrapper, store } = createWrapper({ modules: { level: levelClone, core: coreClone } });

      store.commit('core/SET_STATUS', 'playing');
      expect(wrapper.vm.$refs.canvas).toMatchObject({ width: 160, height: 120 });
    });

    describe('при изменении stateChange', () => {
      it('должен отрисовать анимацию на canvas', () => {
        const levelClone = clone(level);
        const coreClone = clone(core);
        const lava = createLava(createVector(0, 0), '=');
        const player = createPlayer(createVector(2, 1));
        const coin = createCoin(createVector(3, 0.5), createVector(3, 0.5), Math.PI);
        const time = 0.12;

        levelClone.state.levelHeight = 3;
        levelClone.state.levelWidth = 4;
        levelClone.state.rows = [
          ['empty', 'empty', 'empty', 'empty'],
          ['empty', 'empty', 'empty', 'wall'],
          ['lava', 'wall', 'wall', 'wall'],
        ];
        coreClone.state.actors = [lava, player, coin];
        const { wrapper, store } = createWrapper({
          modules: {
            level: levelClone,
            core: coreClone,
          },
        });

        store.commit('core/SET_ACTORS', [
          lava.update(time, () => true),
          player.update(time, () => true, { ArrowLeft: true }),
          coin.update(time, () => true),
        ]);
        store.commit('core/SET_STATUS', 'playing');
        checkCanvas(wrapper);
      });

      it(`должен обновить положение видимой части canvas при движении игрока
        вверх и вправо`, async () => {
        const levelClone = clone(level);
        const coreClone = clone(core);

        levelClone.state.levelHeight = 30;
        levelClone.state.levelWidth = 30;
        levelClone.state.rows = Array.from(
          { length: 30 },
          (index) => {
            const entity = index === 8 ? 'wall' : 'empty';

            return Array.from({ length: 30 }, () => entity);
          },
        );
        coreClone.state.actors = [createPlayer(createVector(7, 18.5))];
        const { wrapper, store } = createWrapper({
          modules: {
            level: levelClone,
            core: coreClone,
          },
        });

        store.commit('core/SET_ACTORS', [createPlayer(createVector(14.2, 0.5))]);
        store.commit('core/SET_STATUS', 'playing');
        expect(wrapper.vm.viewport).toMatchObject({ top: 0, left: 4.600000000000001 });
      });

      it(`должен обновить положение видимой части canvas при движении игрока
        вниз и влево`, async () => {
        const levelClone = clone(level);
        const coreClone = clone(core);

        levelClone.state.levelHeight = 30;
        levelClone.state.levelWidth = 30;
        levelClone.state.rows = Array.from(
          { length: 30 },
          (index) => {
            const entity = index === 8 ? 'wall' : 'empty';

            return Array.from({ length: 30 }, () => entity);
          },
        );
        coreClone.state.actors = [createPlayer(createVector(20, 0.5))];
        const { wrapper, store } = createWrapper({
          modules: {
            level: levelClone,
            core: coreClone,
          },
        });

        store.commit('core/SET_ACTORS', [createPlayer(createVector(2.2, 17.5))]);
        store.commit('core/SET_STATUS', 'playing');
        expect(wrapper.vm.viewport).toMatchObject({ top: 11.5, left: 0 });
      });
    });

    it('при изменении countEndLevel должен очистить canvas', () => {
      const levelClone = clone(level);
      const coreClone = clone(core);

      levelClone.state.levelHeight = 3;
      levelClone.state.levelWidth = 4;
      levelClone.state.rows = [
        ['empty', 'empty', 'empty', 'empty'],
        ['empty', 'empty', 'empty', 'wall'],
        ['empty', 'wall', 'wall', 'wall'],
      ];
      coreClone.state.actors = [
        createLava(createVector(0, 0), '='),
        createPlayer(createVector(2, 1)),
      ];
      const { wrapper, store } = createWrapper({ modules: { level: levelClone, core: coreClone } });
      const mock = jest.fn();

      store.commit('core/SET_STATUS', 'playing');
      wrapper.vm.cx.clearRect = mock;
      store.commit('level/ADD_END_LEVEL');
      expect(mock).toBeCalledWith(0, 0, 160, 120);
    });
  });
});
