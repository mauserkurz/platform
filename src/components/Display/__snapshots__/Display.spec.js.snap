// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Display rendering should render html 1`] = `<div class="display"><canvas></canvas> <img src="@/assets/sprites.png" class="display__sprite"> <img src="@/assets/player.png" class="display__sprite"></div>`;

exports[`component Display store interaction при изменении stateChange должен отрисовать анимацию на canvas 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#34a6fb",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 120,
      "width": 160,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 120,
      "dy": 40,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 0,
      "dy": 80,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 40,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 40,
      "dy": 80,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 80,
      "dy": 80,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 120,
      "dy": 80,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 40,
      "dWidth": 40,
      "dx": 0,
      "dy": 0,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 40,
      "sWidth": 40,
      "sx": 40,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  Object {
    "props": Object {
      "x": 96,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      96,
      0,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {
      "x": -1,
      "y": 1,
    },
    "transform": Array [
      -1,
      -0,
      0,
      1,
      96,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "x": -96,
      "y": 0,
    },
    "transform": Array [
      -1,
      -0,
      0,
      1,
      192,
      0,
    ],
    "type": "translate",
  },
  Object {
    "props": Object {
      "dHeight": 60,
      "dWidth": 48,
      "dx": 72,
      "dy": 20,
      "img": <img
        class="display__sprite"
        src="@/assets/player.png"
      />,
      "sHeight": 60,
      "sWidth": 48,
      "sx": 48,
      "sy": 0,
    },
    "transform": Array [
      -1,
      -0,
      0,
      1,
      192,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  Object {
    "props": Object {
      "dHeight": 24,
      "dWidth": 24,
      "dx": 120,
      "dy": 20,
      "img": <img
        class="display__sprite"
        src="@/assets/sprites.png"
      />,
      "sHeight": 24,
      "sWidth": 24,
      "sx": 80,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
